lms/config/resources.py

```diff
@@ -64,6 +64,33 @@ def h_display_name(self):
 
         return display_name
 
+    @property
+    def h_groupid(self):
+        """
+        Return a unique h groupid for the current request.
+
+        The returned ID is suitable for use with the h API's ``groupid`` parameter.
+
+        The groupid is deterministic and is unique to the LTI course. Calling this
+        function again with params representing the same LTI course will always
+        return the same groupid. Calling this function with different params will
+        always return a different groupid.
+
+        :raise HTTPBadRequest: if an LTI param needed for generating the group
```

Comment:
Raising a 400 here gave me pause, though I think it may be Fine and Proper in pyramid land. This is, after all, traversal time, and you won't have been able to have validated the request/payload yet, so I suppose this makes sense? Just kinda thinking it through...

Reply:
This 400 would actually be raised at view time -- this property is on the context object, but (like many context object attributes) nothing tries to read this property during traversal, it's only called by the view.

Still, I think it _is_ valid to raise a 404 during traversal and that will just 404 the request. I wouldn't be surprised if I'm wrong about that though! Anyway, thankfully it doesn't do that

Reply:
Re validation ... When we get round to having a centralised `lms.validation` package then the view will call that first before doing anything else. At that point these context object properties will be able to assume that the params exist, and the `raise HTTPBadRequest`s here can go